### PR TITLE
Ndf Controllers

### PR DESCRIPTION
**Pull request by utkarsh530**
As said in #1130, the order and step changes for NDF are moved to `controllers.jl`. Please review.

1. I assumed `qold` to be change in `dt` for failed step-size by looking at other implementations. Am I correct?
2. There is some redundant logic I think. First, we calculate q for `stepsize_controller`. Then we check val `integrator.accept_step` and call `step_accept_controller` which checks for some edge cases of `q`. In this implementation, how update the order? should I move it to step_accept_controller and `step_reject_controller` or current `QNDF_stepsize_and_order!` checking conditions for `integrator.EEst <=1.0` is fine?
3. What to put in the controller field in `QNDF` struct [here](https://github.com/SciML/OrdinaryDiffEq.jl/blob/master/src/algorithms.jl#L552-L557)?
I made a separate PR because further moving with diagnosing the problem, overhauling was needed. Your views are highly appreciated.
4. Is the implementation complete? Some optimisations required?
@ChrisRackauckas @kanav99 please review.

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
>Is the implementation complete? Some optimisations required?

Check to see if the allocations are increasing over time when `save_everystep=false`. They shouldn't so this is a good check to see if anything in the inner loop is allocating.

**Comment by utkarsh530** [comment]
This is the result:
Before:
```
julia> function test!(du,u,p,t)
            du[1] = -10*u[1]
            du[2] = -0.01*u[2]
       end
test! (generic function with 1 method)

julia> u0 = [1.01;1.01]
2-element Array{Float64,1}:
 1.01
 1.01

julia> tspan = (0.0, 1.0)
(0.0, 1.0)

julia> prob = ODEProblem(test!, u0, tspan)
ODEProblem with uType Array{Float64,1} and tType Float64. In-place: true
timespan: (0.0, 1.0)
u0: [1.01, 1.01]

julia> sol = solve(prob, QNDF())
retcode: Success
Interpolation: 3rd order Hermite
t: 73-element Array{Float64,1}:
 0.0
 1.4156137738202872e-7
 2.8312275476405744e-7
 4.2468413214608615e-7
 ⋮
 0.8675199816377538
 0.9133884906285924
 0.9788926234660211
 1.0

u: 73-element Array{Array{Float64,1},1}:
 [1.01, 1.01]
 [1.0099985702321124, 1.0099999985702302]
 [1.0099971404664203, 1.0099999971404603]
 [1.0099957107020594, 1.0099999957106904]
 ⋮
 [0.00019415788672002296, 1.0012761953275475]
 [0.00012109367164449898, 1.0008170258133373]
 [6.149948015680113e-5, 1.0001616531997846]  
 [5.066347431470918e-5, 0.9999505720605271] 

```
After:
```
julia> sol = solve(prob, QNDF(),save_everystep=false)
retcode: Success
Interpolation: 1st order linear
t: 2-element Array{Float64,1}:
 0.0
 1.0
u: 2-element Array{Array{Float64,1},1}:
 [1.01, 1.01]
 [5.066347431470918e-5, 0.9999505720605271]
```

**Comment by ChrisRackauckas** [comment]
No I mean, take an example problem, solve it short and solve it long both with `save_everystep=false`. Put `@time` on them: are the allocations different?

**Comment by kanav99** [comment]
> I assumed qold to be change in dt for failed step-size by looking at other implementations.

Try to find that out yourself, which variable changes where. You will get used to codebase more.

> What to put in the controller field in QNDF struct here?

Make it default to something like `:QNDF`.

> In this implementation, how update the order?

What do you mean? You would need to check the order conditions and change it? should go in the controller code where you accept/reject the step

File: src/integrators/controllers.jl [hunk 1]
```diff
@@ -136,15 +136,139 @@ end
 
 function stepsize_controller!(integrator, alg::QNDF)
   cnt = integrator.iter
+  EEst1 = integrator.cache.EEst1
+  EEst2 = integrator.cache.EEst2
   if cnt <= 3
-    q = standard_stepsize_controller!(integrator, alg)
+    # q = standard_stepsize_controller!(integrator, alg)
+    q = 1 #quasi-contsant steps
```
Comment (kanav99) [inline src/integrators/controllers.jl:143]:
```suggestion
    q = one(integrator.qold) #quasi-contsant steps    
```

File: src/integrators/controllers.jl [hunk 2]
```diff
@@ -136,15 +136,139 @@ end
...
+    dt_optim_success, dt_optim_failed = QNDF_stepsize_and_order!(integrator.cache, integrator.EEst, EEst1, EEst2, integrator.dt, integrator.cache.order)
+    q = integrator.dt/dt_optim_success
+    integrator.qold = integrator.dt/dt_optim_failed
   end
   q
 end
 
+function step_accept_controller!(integrator,alg::QNDF,q)
+  #step is accepted, reset count of consecutive failed steps
+  integrator.cache.c = 0
```
Comment (kanav99) [inline src/integrators/controllers.jl:154]: we should have a better variable name for number of failed steps. It's fine if the paper uses `c` and we can make this change later

**Comment by utkarsh530** [comment]
@ChrisRackauckas I tried this:
Before:
```
julia> for tfinal = 1.0:2.0:10.0
           g(u, p, t) = 1.01 * u
           u0 = 1 / 2
           tspan = (0.0, tfinal)
           prob = ODEProblem(g, u0, tspan)
           integrator = init(prob, QNDF(), save_everystep = false)
           @time solve!(integrator)
           end
  0.669273 seconds (1.37 M allocations: 69.249 MiB, 4.46% gc time)
  0.000671 seconds (382 allocations: 82.891 KiB)
  0.001149 seconds (635 allocations: 138.406 KiB)
  0.001062 seconds (893 allocations: 195.000 KiB)
  0.001674 seconds (1.14 k allocations: 248.531 KiB)
```
After:
```
julia> for tfinal = 1.0:2.0:10.0
           g(u, p, t) = 1.01 * u
           u0 = 1 / 2
           tspan = (0.0, tfinal)
           prob = ODEProblem(g, u0, tspan)
           integrator = init(prob, QNDF(), save_everystep = false)
           @time solve!(integrator)
           end
  0.651169 seconds (1.25 M allocations: 63.755 MiB, 5.58% gc time)
  0.000375 seconds (142 allocations: 28.734 KiB)
  0.000206 seconds (199 allocations: 40.719 KiB)
  0.000342 seconds (232 allocations: 47.047 KiB)
  0.000460 seconds (306 allocations: 61.953 KiB)
```

**Comment by ChrisRackauckas** [comment]
yeah much better but still not great. We can track that down in another PR, but the allocations shouldn't be growing.

**Comment by utkarsh530** [comment]
@ChrisRackauckas thanks for the comments! What should be the workflow? Should I first fix this or work on getting correct adaptivity? I checked from previous implementation the results are same. Should work on this controllers implementation (is it mergeable too, or leave as it is for now)?

File: src/integrators/controllers.jl (hunk 2 above)
Comment (utkarsh530) [reply src/integrators/controllers.jl]: Made the requested changes.

**Comment by ChrisRackauckas** [comment]
Rebase onto master which fixed up the tests.

**Comment by utkarsh530** [comment]
@ChrisRackauckas @kanav99 I have made some changes to get better results. Here are the results:
```
julia> for tfinal = 1.0:2.0:60.0
           g(u, p, t) = 1.01 * u
           u0 = 1 / 2
           tspan = (0.0, tfinal)
           prob = ODEProblem(g, u0, tspan)
           integrator = init(prob, QNDF(), save_everystep = false)
           @time solve!(integrator)
       end
  0.414468 seconds (1.16 M allocations: 58.856 MiB, 4.72% gc time)
  0.000051 seconds (50 allocations: 12.922 KiB)
  0.000133 seconds (70 allocations: 18.547 KiB)
  0.000144 seconds (76 allocations: 20.234 KiB)
  0.000128 seconds (96 allocations: 25.859 KiB)
  0.000124 seconds (105 allocations: 28.391 KiB)
  0.000241 seconds (105 allocations: 28.391 KiB)
  0.000526 seconds (105 allocations: 28.391 KiB)
  0.000187 seconds (105 allocations: 28.391 KiB)
  0.000175 seconds (105 allocations: 28.391 KiB)
  0.000316 seconds (105 allocations: 28.391 KiB)
  0.000292 seconds (105 allocations: 28.391 KiB)
  0.000299 seconds (105 allocations: 28.391 KiB)
  0.000308 seconds (105 allocations: 28.391 KiB)
  0.000300 seconds (105 allocations: 28.391 KiB)
  0.000265 seconds (105 allocations: 28.391 KiB)
  0.000421 seconds (105 allocations: 28.391 KiB)
  0.000424 seconds (105 allocations: 28.391 KiB)
  0.000553 seconds (105 allocations: 28.391 KiB)
  0.000315 seconds (105 allocations: 28.391 KiB)
  0.000320 seconds (105 allocations: 28.391 KiB)
  0.000499 seconds (105 allocations: 28.391 KiB)
  0.000455 seconds (105 allocations: 28.391 KiB)
  0.000427 seconds (105 allocations: 28.391 KiB)
  0.000396 seconds (105 allocations: 28.391 KiB)
  0.000396 seconds (105 allocations: 28.391 KiB)
  0.000548 seconds (105 allocations: 28.391 KiB)
  0.000569 seconds (105 allocations: 28.391 KiB)
  0.000470 seconds (105 allocations: 28.391 KiB)
  0.000454 seconds (105 allocations: 28.391 KiB)
```
I think there's some change initially because I am updating it if `cnt < k`, but its effect is not pronounced after (as seen above).

File: src/caches/bdf_caches.jl [hunk 3]
```diff
@@ -296,9 +302,14 @@ function alg_cache(alg::QNDF,u,rate_prototype,uEltypeNoUnits,uBottomEltypeNoUnit
   R = fill(zero(t), 5, 5)
   U = fill(zero(t), 5, 5)
 
+  EEst1 = tTypeNoUnits(1)
+  EEst2 = tTypeNoUnits(1)
+
   max_order = 5
 
-  QNDFConstantCache(nlsolver,D,D2,R,U,1,max_order,udiff,dts,h,0)
+  γₖ = [sum(1//j for j in 1:k) for k in 1:6]
```
Comment (ChrisRackauckas) [inline src/caches/bdf_caches.jl:310]: in fact, this should probably be a static array.

Comment (ChrisRackauckas) [inline src/caches/bdf_caches.jl:310]:
```suggestion
  γₖ = SA[sum(1//j for j in 1:k) for k in 1:6]
```

See https://github.com/JuliaArrays/StaticArrays.jl

Comment (kanav99) [reply src/caches/bdf_caches.jl:310]: I guess it changes

**Comment by kanav99** [comment]
Much better but I am seeing different results:
```
julia> for tfinal = 1.0:2.0:60.0
                  g(u, p, t) = 1.01 * u
                  u0 = 1 / 2
                  tspan = (0.0, tfinal)
                  prob = ODEProblem(g, u0, tspan)
                  integrator = init(prob, QNDF(), save_everystep = false)
                  @time solve!(integrator)
              end
  1.404070 seconds (5.29 M allocations: 256.201 MiB, 4.70% gc time)
  0.000048 seconds (50 allocations: 12.922 KiB)
  0.000068 seconds (70 allocations: 18.547 KiB)
  0.000095 seconds (76 allocations: 20.234 KiB)
  0.000078 seconds (80 allocations: 21.359 KiB)
  0.000072 seconds (80 allocations: 21.359 KiB)
  0.000076 seconds (80 allocations: 21.359 KiB)
  0.000083 seconds (80 allocations: 21.359 KiB)
  0.000093 seconds (89 allocations: 23.891 KiB)
  0.000105 seconds (102 allocations: 27.547 KiB)
  0.000117 seconds (107 allocations: 28.953 KiB)
  0.000146 seconds (156 allocations: 42.734 KiB)
  0.000156 seconds (162 allocations: 44.422 KiB)
  0.000168 seconds (169 allocations: 46.391 KiB)
  0.000171 seconds (169 allocations: 46.391 KiB)
  0.000191 seconds (194 allocations: 53.422 KiB)
  0.000211 seconds (224 allocations: 61.859 KiB)
  0.000351 seconds (230 allocations: 63.547 KiB)
  0.000238 seconds (242 allocations: 66.922 KiB)
  0.000250 seconds (249 allocations: 68.891 KiB)
  0.000267 seconds (272 allocations: 75.359 KiB)
  0.000272 seconds (272 allocations: 75.359 KiB)
  0.000279 seconds (272 allocations: 75.359 KiB)
  0.000306 seconds (301 allocations: 83.516 KiB)
  0.000318 seconds (307 allocations: 85.203 KiB)
  0.000342 seconds (356 allocations: 98.984 KiB)
  0.000372 seconds (362 allocations: 100.672 KiB)
  0.000363 seconds (367 allocations: 102.078 KiB)
  0.000370 seconds (367 allocations: 102.078 KiB)
  0.000402 seconds (367 allocations: 102.078 KiB)```

**Comment by utkarsh530** [comment]
@kanav99 that's weird. I updated Julia version too the results are the same. Let me test it on another machine
These are the current results:
```
julia> for tfinal = 1.0:2.0:60.0
           g(u, p, t) = 1.01 * u
           u0 = 1 / 2
           tspan = (0.0, tfinal)
           prob = ODEProblem(g, u0, tspan)
           integrator = init(prob, QNDF(), save_everystep = false)
           @time solve!(integrator)
       end
  0.494513 seconds (1.17 M allocations: 59.367 MiB, 5.26% gc time)
  0.000076 seconds (50 allocations: 12.922 KiB)
  0.000138 seconds (70 allocations: 18.547 KiB)
  0.000136 seconds (76 allocations: 20.234 KiB)
  0.000199 seconds (127 allocations: 34.578 KiB)
  0.000159 seconds (127 allocations: 34.578 KiB)
  0.000266 seconds (127 allocations: 34.578 KiB)
  0.000240 seconds (127 allocations: 34.578 KiB)
  0.000404 seconds (127 allocations: 34.578 KiB)
  0.000416 seconds (127 allocations: 34.578 KiB)
  0.000282 seconds (127 allocations: 34.578 KiB)
  0.000645 seconds (127 allocations: 34.578 KiB)
  0.000460 seconds (127 allocations: 34.578 KiB)
  0.000592 seconds (127 allocations: 34.578 KiB)
  0.000738 seconds (127 allocations: 34.578 KiB)
  0.000656 seconds (127 allocations: 34.578 KiB)
  0.000771 seconds (127 allocations: 34.578 KiB)
  0.000549 seconds (127 allocations: 34.578 KiB)
  0.000649 seconds (127 allocations: 34.578 KiB)
  0.000553 seconds (127 allocations: 34.578 KiB)
  0.001046 seconds (127 allocations: 34.578 KiB)
  0.000689 seconds (127 allocations: 34.578 KiB)
  0.000698 seconds (127 allocations: 34.578 KiB)
  0.000747 seconds (127 allocations: 34.578 KiB)
  0.000605 seconds (127 allocations: 34.578 KiB)
  0.000425 seconds (127 allocations: 34.578 KiB)
  0.000578 seconds (127 allocations: 34.578 KiB)
  0.000995 seconds (127 allocations: 34.578 KiB)
  0.001241 seconds (127 allocations: 34.578 KiB)
  0.000748 seconds (127 allocations: 34.578 KiB)
```

**Comment by utkarsh530** [comment]
Hi @ChrisRackauckas and @kanav99, I made some improvements in OOP for QNDF, please have a look and review:
[This](https://gist.github.com/utkarsh530/02489370057c4068416eb4fca334dd02) is the script I am using for checking allocations with `--track-allocation=user`
Results:
```
julia> include("test_allocations.jl")
  0.432716 seconds (703.66 k allocations: 36.104 MiB, 4.48% gc time)
  0.001234 seconds (1 allocation: 112 bytes)
  0.002133 seconds (1 allocation: 112 bytes)
  0.002981 seconds (1 allocation: 112 bytes)
  0.003632 seconds (1 allocation: 112 bytes)
  0.004049 seconds (1 allocation: 112 bytes)
  0.005287 seconds (1 allocation: 112 bytes)
  0.005348 seconds (1 allocation: 112 bytes)
  0.005432 seconds (1 allocation: 112 bytes)
  0.006640 seconds (1 allocation: 112 bytes)
  0.007436 seconds (1 allocation: 112 bytes)
  0.008043 seconds (1 allocation: 112 bytes)
  0.008405 seconds (1 allocation: 112 bytes)
  0.009014 seconds (1 allocation: 112 bytes)
  0.009399 seconds (1 allocation: 112 bytes)
  0.000377 seconds (1 allocation: 112 bytes)
  0.002210 seconds (1 allocation: 112 bytes)
  0.002822 seconds (1 allocation: 112 bytes)
  0.002932 seconds (1 allocation: 112 bytes)
  0.005228 seconds (1 allocation: 112 bytes)
  0.005266 seconds (1 allocation: 112 bytes)
  0.006337 seconds (1 allocation: 112 bytes)
  0.006453 seconds (1 allocation: 112 bytes)
  0.007132 seconds (1 allocation: 112 bytes)
  0.007199 seconds (1 allocation: 112 bytes)
  0.007552 seconds (1 allocation: 112 bytes)
  0.008506 seconds (1 allocation: 112 bytes)
  0.009428 seconds (1 allocation: 112 bytes)
  0.009161 seconds (1 allocation: 112 bytes)
  0.010230 seconds (1 allocation: 112 bytes)
```
Fixing OOP for now.

**Comment by ChrisRackauckas** [comment]
Did it get slower?

**Comment by utkarsh530** [comment]
```
julia> include("test_allocations_oop.jl")
[ Info: Precompiling OrdinaryDiffEq [1dea7af3-3e70-54e6-95c3-0bf5283fa5ed]
  3.771898 seconds (4.59 M allocations: 223.647 MiB, 7.81% gc time)
  0.002329 seconds (46 allocations: 12.766 KiB)
  0.004516 seconds (66 allocations: 18.391 KiB)
  0.003909 seconds (72 allocations: 20.078 KiB)
  0.006004 seconds (123 allocations: 34.422 KiB)
  0.006633 seconds (123 allocations: 34.422 KiB)
  0.008929 seconds (123 allocations: 34.422 KiB)
  0.006489 seconds (123 allocations: 34.422 KiB)
  0.007084 seconds (123 allocations: 34.422 KiB)
  0.007599 seconds (123 allocations: 34.422 KiB)
  0.008123 seconds (123 allocations: 34.422 KiB)
  0.008284 seconds (123 allocations: 34.422 KiB)
  0.008524 seconds (123 allocations: 34.422 KiB)
  0.008835 seconds (123 allocations: 34.422 KiB)
  0.010547 seconds (123 allocations: 34.422 KiB)
  0.000433 seconds (10 allocations: 2.641 KiB)
  0.002902 seconds (46 allocations: 12.766 KiB)
  0.003413 seconds (66 allocations: 18.391 KiB)
  0.003787 seconds (72 allocations: 20.078 KiB)
  0.005017 seconds (123 allocations: 34.422 KiB)
  0.005289 seconds (123 allocations: 34.422 KiB)
  0.005680 seconds (123 allocations: 34.422 KiB)
  0.006644 seconds (123 allocations: 34.422 KiB)
  0.007061 seconds (123 allocations: 34.422 KiB)
  0.007323 seconds (123 allocations: 34.422 KiB)
  0.007267 seconds (123 allocations: 34.422 KiB)
  0.008028 seconds (123 allocations: 34.422 KiB)
  0.009367 seconds (123 allocations: 34.422 KiB)
  0.008762 seconds (123 allocations: 34.422 KiB)
  0.009381 seconds (123 allocations: 34.422 KiB)
```
Not that much.. I just ran this on two commits before..

**Comment by ChrisRackauckas** [comment]
looks like it scales better though. Good

**Comment by utkarsh530** [comment]
Thanks! A general doubt though, can't we optimise `R =R*U` through any LinearAlg functions like `mul!` (which does not permit this) without allocating another matrix?

**Comment by ChrisRackauckas** [comment]
`mul!(R,R,U)` is fine IIRC

**Comment by utkarsh530** [comment]
This is the error log I got:

```
julia> include("test_allocations_oop.jl")
[ Info: Precompiling OrdinaryDiffEq [1dea7af3-3e70-54e6-95c3-0bf5283fa5ed]
ERROR: LoadError: ArgumentError: output matrix must not be aliased with input matrix
Stacktrace:
 [1] gemm_wrapper!(::Array{Float64,2}, ::Char, ::Char, ::Array{Float64,2}, ::Array{Float64,2}, ::LinearAlgebra.MulAddMul{true,true,Float64,Float64}) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.3\LinearAlgebra\src\matmul.jl:549
 [2] mul! at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.3\LinearAlgebra\src\matmul.jl:160 [inlined]
 [3] mul! at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.3\LinearAlgebra\src\matmul.jl:203 [inlined]
 [4] perform_step!(::OrdinaryDiffEq.ODEIntegrator{QNDF{0,true,DefaultLinSolve,NLNewton{Rational{Int64},Rational{Int64},Rational{Int64}},DataType,Nothing,Nothing,NTuple{5,Float64}},false,Float64,Nothing,Float64,DiffEqBase.NullParameters,Float64,Float64,Float64,Array{Float64,1},ODESolution{Float64,1,Array{Float64,1},Nothing,Nothing,Array{Float64,1},Array{Array{Float64,1},1},ODEProblem{Float64,Tuple{Float64,Float64},false,DiffEqBase.NullParameters,ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},DiffEqBase.StandardODEProblem},QNDF{0,true,DefaultLinSolve,NLNewton{Rational{Int64},Rational{Int64},Rational{Int64}},DataType,Nothing,Nothing,NTuple{5,Float64}},OrdinaryDiffEq.InterpolationData{ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Array{Float64,1},Array{Float64,1},Array{Array{Float64,1},1},OrdinaryDiffEq.QNDFConstantCache{OrdinaryDiffEq.NLSolver{NLNewton{Rational{Int64},Rational{Int64},Rational{Int64}},false,Float64,Float64,OrdinaryDiffEq.NLNewtonConstantCache{Float64,Float64,Float64,Float64,DiffEqBase.UDerivativeWrapper{ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Float64,DiffEqBase.NullParameters}}},Array{Float64,2},Array{Float64,2},Array{Float64,2},Array{Float64,2},Float64,Array{Float64,2},Float64,Array{Float64,1}}},DiffEqBase.DEStats},ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},OrdinaryDiffEq.QNDFConstantCache{OrdinaryDiffEq.NLSolver{NLNewton{Rational{Int64},Rational{Int64},Rational{Int64}},false,Float64,Float64,OrdinaryDiffEq.NLNewtonConstantCache{Float64,Float64,Float64,Float64,DiffEqBase.UDerivativeWrapper{ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Float64,DiffEqBase.NullParameters}}},Array{Float64,2},Array{Float64,2},Array{Float64,2},Array{Float64,2},Float64,Array{Float64,2},Float64,Array{Float64,1}},OrdinaryDiffEq.DEOptions{Float64,Float64,Float64,Float64,typeof(DiffEqBase.ODE_DEFAULT_NORM),typeof(LinearAlgebra.opnorm),CallbackSet{Tuple{},Tuple{}},typeof(DiffEqBase.ODE_DEFAULT_ISOUTOFDOMAIN),typeof(DiffEqBase.ODE_DEFAULT_PROG_MESSAGE),typeof(DiffEqBase.ODE_DEFAULT_UNSTABLE_CHECK),DataStructures.BinaryHeap{Float64,DataStructures.LessThan},DataStructures.BinaryHeap{Float64,DataStructures.LessThan},Nothing,Nothing,Int64,Array{Float64,1},Array{Float64,1},Array{Float64,1}},Float64,Float64,Nothing,OrdinaryDiffEq.DefaultInit}, ::OrdinaryDiffEq.QNDFConstantCache{OrdinaryDiffEq.NLSolver{NLNewton{Rational{Int64},Rational{Int64},Rational{Int64}},false,Float64,Float64,OrdinaryDiffEq.NLNewtonConstantCache{Float64,Float64,Float64,Float64,DiffEqBase.UDerivativeWrapper{ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Float64,DiffEqBase.NullParameters}}},Array{Float64,2},Array{Float64,2},Array{Float64,2},Array{Float64,2},Float64,Array{Float64,2},Float64,Array{Float64,1}}, ::Bool) at C:\Users\Utkarsh\.julia\dev\OrdinaryDiffEq\src\perform_step\bdf_perform_step.jl:642
 [5] perform_step! at C:\Users\Utkarsh\.julia\dev\OrdinaryDiffEq\src\perform_step\bdf_perform_step.jl:614 [inlined]
 [6] solve!(::OrdinaryDiffEq.ODEIntegrator{QNDF{0,true,DefaultLinSolve,NLNewton{Rational{Int64},Rational{Int64},Rational{Int64}},DataType,Nothing,Nothing,NTuple{5,Float64}},false,Float64,Nothing,Float64,DiffEqBase.NullParameters,Float64,Float64,Float64,Array{Float64,1},ODESolution{Float64,1,Array{Float64,1},Nothing,Nothing,Array{Float64,1},Array{Array{Float64,1},1},ODEProblem{Float64,Tuple{Float64,Float64},false,DiffEqBase.NullParameters,ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},DiffEqBase.StandardODEProblem},QNDF{0,true,DefaultLinSolve,NLNewton{Rational{Int64},Rational{Int64},Rational{Int64}},DataType,Nothing,Nothing,NTuple{5,Float64}},OrdinaryDiffEq.InterpolationData{ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Array{Float64,1},Array{Float64,1},Array{Array{Float64,1},1},OrdinaryDiffEq.QNDFConstantCache{OrdinaryDiffEq.NLSolver{NLNewton{Rational{Int64},Rational{Int64},Rational{Int64}},false,Float64,Float64,OrdinaryDiffEq.NLNewtonConstantCache{Float64,Float64,Float64,Float64,DiffEqBase.UDerivativeWrapper{ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Float64,DiffEqBase.NullParameters}}},Array{Float64,2},Array{Float64,2},Array{Float64,2},Array{Float64,2},Float64,Array{Float64,2},Float64,Array{Float64,1}}},DiffEqBase.DEStats},ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},OrdinaryDiffEq.QNDFConstantCache{OrdinaryDiffEq.NLSolver{NLNewton{Rational{Int64},Rational{Int64},Rational{Int64}},false,Float64,Float64,OrdinaryDiffEq.NLNewtonConstantCache{Float64,Float64,Float64,Float64,DiffEqBase.UDerivativeWrapper{ODEFunction{false,typeof(test),LinearAlgebra.UniformScaling{Bool},Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing,Nothing},Float64,DiffEqBase.NullParameters}}},Array{Float64,2},Array{Float64,2},Array{Float64,2},Array{Float64,2},Float64,Array{Float64,2},Float64,Array{Float64,1}},OrdinaryDiffEq.DEOptions{Float64,Float64,Float64,Float64,typeof(DiffEqBase.ODE_DEFAULT_NORM),typeof(LinearAlgebra.opnorm),CallbackSet{Tuple{},Tuple{}},typeof(DiffEqBase.ODE_DEFAULT_ISOUTOFDOMAIN),typeof(DiffEqBase.ODE_DEFAULT_PROG_MESSAGE),typeof(DiffEqBase.ODE_DEFAULT_UNSTABLE_CHECK),DataStructures.BinaryHeap{Float64,DataStructures.LessThan},DataStructures.BinaryHeap{Float64,DataStructures.LessThan},Nothing,Nothing,Int64,Array{Float64,1},Array{Float64,1},Array{Float64,1}},Float64,Float64,Nothing,OrdinaryDiffEq.DefaultInit}) at C:\Users\Utkarsh\.julia\dev\OrdinaryDiffEq\src\solve.jl:424
 [7] macro expansion at .\util.jl:155 [inlined]
 [8] top-level scope at C:\Users\Utkarsh\Documents\test_allocations_oop.jl:15
 [9] include at .\boot.jl:328 [inlined]
 [10] include_relative(::Module, ::String) at .\loading.jl:1105
 [11] include(::Module, ::String) at .\Base.jl:31
 [12] include(::String) at .\client.jl:424
 [13] top-level scope at REPL[2]:1
in expression starting at C:\Users\Utkarsh\Documents\test_allocations_oop.jl:11
```
I changed `mul!(cache.tmp,R,U)` -> `mul!(R,R,U)` and corresponding change in `reinterpolate_history!`

**Comment by ChrisRackauckas** [comment]
Oh okay, so yeah you can't alias. No worries.

File: src/bdf_utils.jl [hunk 4]
```diff
@@ -52,7 +52,7 @@ function reinterpolate_history!(cache::OrdinaryDiffEqMutableCache, D, R, k)
   fill!(tmp,zero(eltype(D[1])))
   for j = 1:k
     for k = 1:k
-      @.. tmp += D[k] * R[k,j]
```
Comment (ChrisRackauckas) [inline src/bdf_utils.jl:55]: yup it's unsafe with aliasing, good catch

**Comment by ChrisRackauckas** [comment]
There were no conflicts so I rebased and 🤞 it should work

**Comment by utkarsh530** [comment]
I rebased too 😅 , let us wait for the CI build

**Comment by utkarsh530** [comment]
@ChrisRackauckas anything left to do before merging into master?

**Pull request merged**